### PR TITLE
[WIP] Custom root certificates for JWKSResolver (#15086)

### PR DESCRIPTION
**Pull request by dgn**
This is just a POC solution to #15086, there are still a few open points:
- should the PEM come through a Kubernetes secret or the MeshConfig? This uses MeshConfig
- no tests yet (depends on where PEM will live)
- points to my fork of `api` for the `MeshConfig` changes, this is of course also temporary

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: vendor/istio.io/api/mesh/v1alpha1/config.proto [hunk 1]
```diff
@@ -306,8 +306,14 @@ message MeshConfig {
   // It can be enabled by destination using the destinationRule.trafficPolicy.connectionPool.http.h2UpgradePolicy override.
   H2UpgradePolicy h2_upgrade_policy = 41;
```
Comment (quanjielin) [inline vendor/istio.io/api/mesh/v1alpha1/config.proto]: pls update istio/api first (https://github.com/istio/api/blob/master/mesh/v1alpha1/config.proto), then update sha here; btw could we use env variable here ?(it's not recommended to update meshconfig /cc @costinm )

File: pilot/pkg/model/jwks_resolver.go [hunk 2]
```diff
@@ -138,28 +138,40 @@ func NewJwksResolver(evictionDuration, refreshInterval time.Duration) *JwksResol
...
 	atomic.StoreUint64(&ret.refreshJobFetchFailedCount, 0)
 	go ret.refresher()
 
 	return ret
 }
 
+// InitializeSecureClient sets up an HTTPS client to be used for resolving jwks_uri
+func (r *JwksResolver) InitializeSecureClient(usePublicRootCABundle bool, customRootCA string) {
+	caCertPool := x509.NewCertPool()
+	if usePublicRootCABundle {
```
Comment (quanjielin) [inline pilot/pkg/model/jwks_resolver.go]: could we just put public ca bundle by default ?

**Comment by yangminzhu** [comment]
Would you mind to create a doc including the different options and present it in the config and security WG? Thanks.

**Comment by dgn** [comment]
> Would you mind to create a doc including the different options and present it in the config and security WG? Thanks.

Yes, will do that. Thank you

File: pilot/pkg/model/jwks_resolver.go (hunk 2 above)
Comment (dgn) [reply pilot/pkg/model/jwks_resolver.go]: Done.

File: vendor/istio.io/api/mesh/v1alpha1/config.proto (hunk 1 above)
Comment (dgn) [reply vendor/istio.io/api/mesh/v1alpha1/config.proto]: >pls update istio/api first then update sha here

that's how I did it; I pulled in my changes using `GO111MODULE=on go mod vendor` after changing `go.mod`

>btw could we use env variable here ?

that's what I was not sure about. this version is changing the MeshConfig, I am working on a document outlining this plus another version that doesn't change the MeshConfig

**Comment by costinm** [comment]
This is one of the cases where env is not a good choice.

This is clearly not a secret - we are discussing moving the ca root(s) to a
config as well.

Let's discuss this in a doc - Phil is also working on a similar problem for
configuring  root ca for multi cluster and root ca rotation.

Most important is to figure out if it is a global setting or per namespace.
Would all tenants/teams use same roots ?
Second q is if other restrictions will be needed, like limit a public key
to specific ids or namespaces.
And finally would be good to have a consistent api/model for cert and jwt
pubkeys.

**Comment by dgn** [comment]
Dropping this in favor of a simpler solution I am working on

**Pull request closed without merging**